pkg/asset/installconfig/aws/validation.go

```diff
@@ -0,0 +1,162 @@
+	for idx, id := range subnets {
+		if _, ok := publicSubnets[id]; ok {
+			publicSubnetsIdx[id] = idx
+		}
+	}
+
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, privateSubnets, privateSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, publicSubnets, publicSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, privateSubnets, privateSubnetsIdx, "private")...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, publicSubnets, publicSubnetsIdx, "public")...)
+
+	privateZones := sets.NewString()
+	publicZones := sets.NewString()
+	for _, subnet := range privateSubnets {
+		privateZones.Insert(subnet.Zone)
+	}
+	for _, subnet := range publicSubnets {
+		publicZones.Insert(subnet.Zone)
+	}
+	if publish == types.ExternalPublishingStrategy && !publicZones.IsSuperset(privateZones) {
+		errMsg := fmt.Sprintf("No public subnet provided for zones %s", privateZones.Difference(publicZones).List())
+		allErrs = append(allErrs, field.Invalid(fldPath, subnets, errMsg))
+	}
+
+	return allErrs
+}
+
+func validateMachinePool(ctx context.Context, meta *Metadata, fldPath *field.Path, platform *awstypes.Platform, pool *awstypes.MachinePool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(pool.Zones) > 0 {
+		availableZones := sets.String{}
+		if len(platform.Subnets) > 0 {
+			privateSubnets, err := meta.PrivateSubnets(ctx)
+			if err != nil {
+				return append(allErrs, field.InternalError(fldPath, err))
+			}
+			for _, subnet := range privateSubnets {
+				availableZones.Insert(subnet.Zone)
+			}
+		} else {
+			allzones, err := meta.AvailabilityZones(ctx)
+			if err != nil {
+				return append(allErrs, field.InternalError(fldPath, err))
+			}
+			availableZones.Insert(allzones...)
+		}
+
+		if diff := sets.NewString(pool.Zones...).Difference(availableZones); diff.Len() > 0 {
+			errMsg := fmt.Sprintf("No subnets provided for zones %s", diff.List())
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("zones"), pool.Zones, errMsg))
+		}
+	}
+	return allErrs
+}
+
+func validateSubnetCIDR(fldPath *field.Path, subnets map[string]Subnet, idxMap map[string]int, machineCIDR *ipnet.IPNet) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for id, v := range subnets {
+		fp := fldPath.Index(idxMap[id])
+		cidr, _, err := net.ParseCIDR(v.CIDR)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fp, id, err.Error()))
+			continue
+		}
+		if !machineCIDR.Contains(cidr) {
+			errMsg := fmt.Sprintf("CIDR range %s is outside of the MachineCIDR %s", v.CIDR, machineCIDR)
+			allErrs = append(allErrs, field.Invalid(fp, id, errMsg))
+			continue
+		}
+	}
+	return allErrs
+}
+
+func validateDuplicateSubnetZones(fldPath *field.Path, subnets map[string]Subnet, idxMap map[string]int, typ string) field.ErrorList {
+	var keys []string
+	for id := range subnets {
+		keys = append(keys, id)
+	}
+	sort.Strings(keys)
```

Comment:
@wking commented earlier that the sorting provided stability. I find this a little confusing, so perhaps a short comment might save someone some time in the future 

if i'm understanding correctly the (only) purpose of the sorting is to guarantee that the same error message is presented, otherwise because the order of keys is not guaranteed, the message may contain subnet `foo` in one run and `bar` in another.

Reply:
a lot of go code that use maps for uniqueness uses this key sort method to provide the consistency in access order _because map traveral is ransom in Go_, so i'm not sure if the comment is required.